### PR TITLE
alternatives / variants / whatever

### LINKED ISSUE #39: Implement alternates
```
 ()
  . _ = scoop :: chocolate ( )
  . scoop : # vanilla ( ) # chocolate ( ) # strawberry ( )
  . _ : p = p :: point { x = 3 , y = 4 }
  . p : # point { x : int , y : int }
  . tuple : x => y => z => # pair { x : x , y : y } # triplet { x : x , y : y , z : z }
```

### PR DESCRIPTION
**Pull request by surprisetalk**
@tekknolagi @gregorybchris 

Before I start coding, am I at least using everything correctly in the tests?

### CODE REVIEW

File: scrapscript.py [hunk 1]
```diff
@@ -1670,6 +1670,44 @@ def test_parse_record_multiple_fields(self) -> None:
...
+                    Operator("#"), Name("chocolate"), LeftParen(), RightParen(),
+                    Operator("#"), Name("strawberry"), LeftParen(), RightParen(),
+                ]
+            ),
+            Alts({ "vanilla": Hole(), "chocolate": Hole(), "strawberry": Hole() })
+        )
+
+    def test_parse_alt(self) -> None:
+        self.assertEqual(
+            parse([ Operator("#"), Name("a"), Name("int"), Operator("::"), Name("a"), NumLit(8)]),
```
Comment (tekknolagi) [inline scrapscript.py:1707]: Do you want an equals or colon here instead of `a int`?

Comment (tekknolagi) [reply scrapscript.py:1707]: Also with the `a 8`

File: scrapscript.py [hunk 2]
```diff
@@ -1873,6 +1910,33 @@ def test_match_list_with_non_matching_const_returns_none(self) -> None:
...
+    def test_match_alts_returns_none(self) -> None:
+        self.assertEqual(
+            match(
+                Alt("a", Var("x")),
+                pattern=Alts({ "a": Var("int"), "b" : Var("int") }, "b", Int(123)),
+            ),
+            None,
+        )
+
+    def test_match_alts_returns_var(self) -> None:
```
Comment (tekknolagi) [inline scrapscript.py:1922]: Maybe I am missing something but how is this case different from the above?

**Review by tekknolagi — COMMENTED** [review]
I think overall this looks like a good first sketch

Would be good to have some end to end tests too

Re: tests:

* Test names should be more specific so it's clear what they are testing
* Would be good to test both nested alternatives and also interaction with other kinds of types
* (How) does recursion work? Like a tree type with leaf/node

**Comment by surprisetalk** [comment]
Closing this attempt. I'll open a new PR after discussing strategy in #39

**Pull request closed without merging**